File: python-agent/suho_agent/ipc/protocol.py
"""IPC Protocol — Python-side message models matching the Rust protocol exactly."""

from __future__ import annotations

from typing import Any, Literal, Optional, Union
from pydantic import BaseModel, Field


# ─── Agent → CLI (Python sends these) ────────────────────────────────────────

class ToolRequestMessage(BaseModel):
    type: Literal["tool_request"] = "tool_request"
    id: str
    task_id: str
    tool: str
    args: dict[str, Any]
    policy_level: str  # SAFE / MODERATE / DANGEROUS / CRITICAL
    policy_reasons: list[str]
    description: str


class PermissionRequestMessage(BaseModel):
    type: Literal["permission_request"] = "permission_request"
    id: str
    task_id: str
    tool: str
    description: str
    command_preview: Optional[str] = None
    level: str
    consequences: list[str] = Field(default_factory=list)


class StatusUpdateMessage(BaseModel):
    type: Literal["status_update"] = "status_update"
    task_id: str
    status: str  # PENDING / PLANNING / EXECUTING / etc.
    message: str
    step: Optional[str] = None
    progress: Optional[float] = None


class StreamChunkMessage(BaseModel):
    type: Literal["stream_chunk"] = "stream_chunk"
    task_id: str
    content: str


class ToolStartedMessage(BaseModel):
    type: Literal["tool_started"] = "tool_started"
    task_id: str
    tool: str
    description: str


class TaskCompleteMessage(BaseModel):
    type: Literal["task_complete"] = "task_complete"
    task_id: str
    summary: str
    files_changed: list[dict[str, Any]] = Field(default_factory=list)
    tool_calls: int = 0
    iterations: int = 0
    token_usage: dict[str, Any] = Field(default_factory=dict)
    duration_ms: int = 0


class TaskFailedMessage(BaseModel):
    type: Literal["task_failed"] = "task_failed"
    task_id: str
    error: str
    recoverable: bool = True
    suggestion: Optional[str] = None


class PlanGeneratedMessage(BaseModel):
    type: Literal["plan_generated"] = "plan_generated"
    task_id: str
    steps: list[dict[str, Any]]


class ThinkingMessage(BaseModel):
    type: Literal["thinking"] = "thinking"
    task_id: str
    content: str


class InfoMessage(BaseModel):
    type: Literal["info"] = "info"
    task_id: str
    message: str


class AgentErrorMessage(BaseModel):
    type: Literal["agent_error"] = "agent_error"
    task_id: str
    error: str
    recoverable: bool = True


class QueryResponseMessage(BaseModel):
    type: Literal["query_response"] = "query_response"
    id: str
    data: Any


# Union of all agent → CLI messages
AgentMessage = Union[
    ToolRequestMessage,
    PermissionRequestMessage,
    StatusUpdateMessage,
    StreamChunkMessage,
    ToolStartedMessage,
    TaskCompleteMessage,
    TaskFailedMessage,
    PlanGeneratedMessage,
    ThinkingMessage,
    InfoMessage,
    AgentErrorMessage,
    QueryResponseMessage,
]


# ─── CLI → Agent (Rust sends these, Python receives) ─────────────────────────

class TaskRequestMessage(BaseModel):
    type: Literal["task_request"] = "task_request"
    id: str
    request: str
    cwd: str
    mode: str  # interactive / autonomous / dry_run / plan_only / ask_only
    max_iterations: int = 30
    timeout_secs: int = 300


class ToolResultMessage(BaseModel):
    type: Literal["tool_result"] = "tool_result"
    id: str
    success: bool
    exit_code: Optional[int] = None
    stdout: str = ""
    stderr: str = ""
    duration_ms: int = 0
    secrets_redacted: int = 0


class ToolDeniedMessage(BaseModel):
    type: Literal["tool_denied"] = "tool_denied"
    id: str
    reason: str
    level: str


class PermissionDecisionMessage(BaseModel):
    type: Literal["permission_decision"] = "permission_decision"
    id: str
    decision: str  # allow_once / allow_session / allow_tool / deny / always_deny


class CancelMessage(BaseModel):
    type: Literal["cancel"] = "cancel"
    task_id: str


class SessionInfoMessage(BaseModel):
    type: Literal["session_info"] = "session_info"
    task_id: str
    config: dict[str, Any]


class ListToolsMessage(BaseModel):
    type: Literal["list_tools"] = "list_tools"
    verbose: bool = False
    category: Optional[str] = None


class ListModelsMessage(BaseModel):
    type: Literal["list_models"] = "list_models"


class HistoryMessage(BaseModel):
    type: Literal["history"] = "history"
    limit: int = 20


class StatusMessage(BaseModel):
    type: Literal["status"] = "status"


class MemoryListMessage(BaseModel):
    type: Literal["memory_list"] = "memory_list"
    limit: int = 20


class MemorySearchMessage(BaseModel):
    type: Literal["memory_search"] = "memory_search"
    query: str


class MemoryDeleteMessage(BaseModel):
    type: Literal["memory_delete"] = "memory_delete"
    id: str


class MemoryClearMessage(BaseModel):
    type: Literal["memory_clear"] = "memory_clear"


class SessionListMessage(BaseModel):
    type: Literal["session_list"] = "session_list"
    limit: int = 20


class ResumeMessage(BaseModel):
    type: Literal["resume"] = "resume"
    id: Optional[str] = None
    cwd: str


# Union of all CLI → agent messages
CliMessage = Union[
    TaskRequestMessage,
    ToolResultMessage,
    ToolDeniedMessage,
    PermissionDecisionMessage,
    CancelMessage,
    SessionInfoMessage,
    ListToolsMessage,
    ListModelsMessage,
    HistoryMessage,
    StatusMessage,
    MemoryListMessage,
    MemorySearchMessage,
    MemoryDeleteMessage,
    MemoryClearMessage,
    SessionListMessage,
    ResumeMessage,
]

# TypeAdapters for Pydantic v2 Union parsing
from pydantic import TypeAdapter

cli_message_adapter: TypeAdapter[CliMessage] = TypeAdapter(CliMessage)
agent_message_adapter: TypeAdapter[AgentMessage] = TypeAdapter(AgentMessage)

File: rust-cli/src/ipc/protocol.rs
//! IPC Protocol — newline-delimited JSON messages between Rust CLI and Python agent.
//!
//! Design principles:
//! - Python side REQUESTS operations via AgentMessage
//! - Rust side DECIDES via GateDecision and sends CliMessage responses
//! - All messages are newline-terminated JSON objects
//! - Message IDs link requests to responses

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

// ─── Direction: CLI → Python ──────────────────────────────────────────────────

/// Messages sent FROM Rust CLI TO Python agent
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum CliMessage {
    /// Initial task request
    TaskRequest {
        id: String,
        request: String,
        cwd: String,
        mode: RunMode,
        max_iterations: u32,
        timeout_secs: u64,
    },

    /// Result of a tool execution (Rust ran it, returning result to Python)
    ToolResult {
        /// Matches the id in AgentMessage::ToolRequest
        id: String,
        success: bool,
        exit_code: Option<i32>,
        stdout: String,
        stderr: String,
        duration_ms: u64,
        /// Redacted secrets report (if any secrets were found and redacted)
        secrets_redacted: u32,
    },

    /// Tool execution denied by security gate
    ToolDenied {
        id: String,
        reason: String,
        level: PolicyLevel,
    },

    /// User permission decision (after interactive prompt)
    PermissionDecision {
        /// Matches the id in AgentMessage::PermissionRequest
        id: String,
        decision: PermissionDecision,
    },

    /// Cancel the active task
    Cancel { task_id: String },

    /// Session/config info (sent after connection)
    SessionInfo {
        task_id: String,
        config: serde_json::Value,
    },

    /// Query messages from CLI to Agent
    ListTools { verbose: bool, category: Option<String> },
    MemoryList { limit: usize },
    MemorySearch { query: String },
    MemoryDelete { id: String },
    MemoryClear,
    Status,
    History { limit: usize },
    ListModels,
    SessionList { limit: usize },
    Resume { id: Option<String>, cwd: String },

    /// List tools request response
    ToolsResponse { tools: Vec<ToolInfo> },

    /// Models list response
    ModelsResponse { models: Vec<ModelInfo> },

    /// History response
    HistoryResponse { entries: Vec<HistoryEntry> },

    /// Status response
    StatusResponse { status: AgentStatus },

    /// Generic oneshot response
    Response { data: serde_json::Value },

    /// Error from Rust side
    RustError { message: String },
}

// ─── Direction: Python → CLI ──────────────────────────────────────────────────

/// Messages sent FROM Python agent TO Rust CLI
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AgentMessage {
    /// Agent requests a tool to be executed
    /// Rust SecurityGate makes the final allow/deny decision
    ToolRequest {
        id: String,
        task_id: String,
        tool: String,
        args: serde_json::Value,
        /// Policy Engine's assessment (advisory — Rust may override)
        policy_level: PolicyLevel,
        policy_reasons: Vec<String>,
        /// Human-readable description of what this will do
        description: String,
    },

    /// Agent requests user permission (for operations that need explicit approval)
    PermissionRequest {
        id: String,
        task_id: String,
        tool: String,
        description: String,
        command_preview: Option<String>,
        level: PolicyLevel,
        /// What will happen if allowed
        consequences: Vec<String>,
    },

    /// Agent status update (for UI display)
    StatusUpdate {
        task_id: String,
        status: TaskStatus,
        message: String,
        step: Option<String>,
        progress: Option<f32>,
    },

    /// Streaming LLM output chunk
    StreamChunk { task_id: String, content: String },

    /// Tool execution started (before result)
    ToolStarted {
        task_id: String,
        tool: String,
        description: String,
    },

    /// Task completed successfully
    TaskComplete {
        task_id: String,
        summary: String,
        files_changed: Vec<FileChange>,
        tool_calls: u32,
        iterations: u32,
        token_usage: TokenUsage,
        duration_ms: u64,
    },

    /// Task failed
    TaskFailed {
        task_id: String,
        error: String,
        recoverable: bool,
        suggestion: Option<String>,
    },

    /// Plan generated (for plan-only mode)
    PlanGenerated {
        task_id: String,
        steps: Vec<PlanStep>,
    },

    /// LLM thinking (shown in verbose mode)
    Thinking { task_id: String, content: String },

    /// General info message
    Info { task_id: String, message: String },

    /// Error from Python side
    AgentError {
        task_id: String,
        error: String,
        recoverable: bool,
    },

    /// Response to CliMessage::ListTools etc.
    QueryResponse {
        id: String,
        data: serde_json::Value,
    },
}

// ─── Shared types ─────────────────────────────────────────────────────────────

/// Execution mode for a task
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RunMode {
    /// Interactive — agent checks in for MODERATE+ operations
    Interactive,
    /// Autonomous — minimal interruption (DANGEROUS/CRITICAL still prompt)
    Autonomous,
    /// Dry run — show planned actions only
    DryRun,
    /// Plan only — generate plan, do not execute
    PlanOnly,
    /// Ask only — pure LLM response, no tool use
    AskOnly,
}

/// Security classification level
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum PolicyLevel {
    Safe,
    Moderate,
    Dangerous,
    Critical,
}

/// User's permission decision
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum PermissionDecision {
    AllowOnce,
    AllowSession,
    AllowTool,
    Deny,
    AlwaysDeny,
}

/// Task lifecycle status
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum TaskStatus {
    Pending,
    Planning,
    Executing,
    WaitingForApproval,
    Verifying,
    Failed,
    Completed,
    Cancelled,
}

/// A file that was changed during a task
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileChange {
    pub path: String,
    pub operation: FileOperation,
    pub diff: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FileOperation {
    Created,
    Modified,
    Deleted,
    Moved { from: String },
}

/// Token usage for a task
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct TokenUsage {
    #[serde(default)]
    pub input_tokens: u64,
    #[serde(default)]
    pub output_tokens: u64,
    #[serde(default)]
    pub total_tokens: u64,
    #[serde(default)]
    pub estimated_cost_usd: Option<f64>,
}

/// A step in a generated plan
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlanStep {
    pub index: u32,
    pub description: String,
    pub tool: Option<String>,
    pub rationale: Option<String>,
    pub risk_level: PolicyLevel,
}

/// Tool metadata for listing
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolInfo {
    pub name: String,
    pub description: String,
    pub category: String,
    pub permission_level: PolicyLevel,
    pub available: bool,
    pub unavailable_reason: Option<String>,
}

/// Model metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelInfo {
    pub name: String,
    pub provider: String,
    pub available: bool,
    pub context_length: Option<usize>,
}

/// History entry
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistoryEntry {
    pub task_id: String,
    pub request: String,
    pub status: TaskStatus,
    pub created_at: DateTime<Utc>,
    pub duration_ms: Option<u64>,
    pub tool_calls: Option<u32>,
}

/// Agent status for status command
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentStatus {
    pub active_task: Option<String>,
    pub model: String,
    pub provider: String,
    pub memory_enabled: bool,
    pub total_tasks_run: u64,
    pub uptime_secs: u64,
}
